pkg/apis/controller/v1alpha1/types.go

```diff
@@ -0,0 +1,128 @@
+/*
+Copyright 2019 The GitLab-Controller Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"fmt"
+	"strconv"
+
+	xpcorev1alpha1 "github.com/crossplaneio/crossplane/pkg/apis/core/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// GitLabSpec defines the desired state of GitLab
+type GitLabSpec struct {
+	// Domain which will contain records to resolve gitlab, registry, and minio (if enabled) to the appropriate IP
+	Domain string `json:"domain"`
+	// HostSuffix appended to domain records, i.e. `gitlab-suffix.domain`, default - no suffix
+	HostSuffix string `json:"hostSuffix,omitempty"`
+	// Protocol http or https, default - https
+	Protocol string `json:"protocol,omitempty"`
+	// Port gitlab service port, default - none
+	Port uint `json:"port,omitempty"`
+	// Email address to register TLS certificates
+	Email string `json:"email"`
+
+	// ProviderRef cloud provider reference
+	ProviderRef corev1.LocalObjectReference `json:"providerRef"`
+	// ReclaimPolicy controls application cleanup
+	ReclaimPolicy xpcorev1alpha1.ReclaimPolicy `json:"reclaimPolicy,omitempty"`
+}
+
+// GitLabStatus defines the observed state of GitLab
+type GitLabStatus struct {
+	xpcorev1alpha1.ConditionedStatus
+
+	// Endpoint for GitLab service
+	Endpoint string `json:"endpoint,omitempty"`
+
+	// State of the GitLab service - string representation of the latest active condition
+	// This is provided for convenience for displaying status gitlab
+	State xpcorev1alpha1.ConditionType `json:"state,omitempty"`
+}
```

Comment:
How will we relate a GitLab CRD back to the `KubernetesApplication` resources it creates and manages so that a user can easily figure out which KubernetesApplications belong to which Gitlab? Perhaps we should use the LabelSelector pattern, simply to document a relationship expectation?

This is missing from my current implementation of the KubernetesApplication -> KubernetesApplicationResource relationship too.

Reply:
I was thinking to use naming convention instead of labels (in addition to `OwnerReference`). 
This is similar to `crossplane-instance`->`managed-resource`, where we use prefix + UID value/
For example: `GitLab {name: foo, uid: bar-uid}` will create `Application {name: gitaly-bar-uid}`, etc.

Reply:
I think that using labels provides a better (or at least more typical) user experience, because `kubectl` and most clients support filtering by label, but not (to my knowledge) by name prefix.

It does seem a little counterintuitive to add a "label selector" that doesn't really "select" anything but instead only documents what label to use to find child resources, but for me at least this is the pattern I'm accustomed to. i.e. Look at Deployment/Service/DaemonSet label selector (or template labels) and use those to find resources via `kubectl get coolstuff -l foo=bar`. For resources that create multiple homogeneous children from one template you could omit the label selector and just look at the template's labels, but that's not the case here.

This all assumes we should design for the case in which a user is interacting with the Crossplane API server via `kubectl`, which may not be the case if we expect these resources to be consumed and managed only via our web UI.

Reply:
I am not opposed to label selector in addition to the naming convention.

Reply:
I feel we may want to add this label selector at some point but it's not important for this first iteration, so I'll resolve this thread.